### ISSUE TITLE
Tree shaking in embroider build with `importSync` and `macroCondition` not working properly

### ISSUE THREAD

**Issue by mydea** [embroider-build/embroider#945]
I am trying to migrate an in-repo addon that I conditionally exclude from the build to regular app code that is statically stripped of my build if not needed in an embroider build.

In an app using embroider to build with all build flags enabled, as well as `staticAppPaths` configured like this:

```js
 return require('@embroider/compat').compatBuild(app, Webpack, {
    staticAddonTestSupportTrees: true,
    staticAddonTrees: true,
    staticHelpers: true,
    staticComponents: true,
   
    staticAppPaths: ['utils'],
});
```

I have an instance initializer:

```js
import ApplicationInstance from '@ember/application/instance';
import { getOwnConfig, importSync, macroCondition } from '@embroider/macros';

export function initialize(appInstance) {
  if (macroCondition(!getOwnConfig().mockGraphqlApi)) {
    return;
  }

   let { setupMockGraphql } = importSync(
    './../utils/mock-graphql/setup-mock-graphql'
  );

  setupMockGraphql(appInstance);
}

export default {
  initialize,
};
```

Which _works_, but it seems that all the code that is used by `utils/mock-graphql/setup-mock-graphql` is _always_ included in the build output (even if it is not actually used). My understanding of `importSync` & `macroCondition` is that it should not be included in that case?

If I remove the `importSync(...)` code, the code is not included in the build output, as it is not referenced anywhere else.

Is this supposed to work that way, or is this a bug in the build-time resolution code?

**Comment by mydea** [comment]
Hey Simon,

finally got around to try this. It seems to actually be working that way! Still _seems_ to me it should work the same both way, but I guess this is then probably more of an issue in how the JS optimization works? As:

```js
function() {
  if (true) {
    return;
  }

  doSomething();
}
```

Should reasonably be optimized to:

```js
function() {}
```

But I guess I at least have a strategy then :sweat_smile: Thanks a lot for the tip!

**Comment by mydea** [comment]
Yeah, that is also what I would have expected - but does not seem to be the case!

**Comment by ef4** [comment]
Oh, so I didn't read this issue carefully enough before. This is a misunderstand of why `macroCondition` exists.

Consider that we could have just not had `macroCondition`. So instead of:

```js
if (macroCondition(getOwnConfig().mockGraphqlApi)) { ... }
```

we could instead let you type:

```js
if (getOwnConfig().mockGraphqlApi) { ... }
```

This has exactly as much information available as the original, so what is the point of `macroCondition`?

The point is that it gives you a *guaranteed* branch elimination point. It creates a build error if we can't eliminate one of the branches, so you know you can rely on it. Other kinds of opportunistic optimization don't give you that guarantee, because they're opportunistic. It's too easy to accidentally lose your optimization without noticing it:

```js
// this would silently lose the branch elimination
if (getOwnConfig().mockGraphqlApi && somethingElse()) { ... }

// this is a build error
if (macroCondition(getOwnConfig().mockGraphqlApi && somethingElse())) { ... }
```
```

In this particular case, production minification probably does eliminate the body of the function but that happens too late in the build process to avoid pulling in the dependency. By the time minification is running, ES module boundaries have been erased, so we can't tell anymore why `setup-mock-graphql` has been appended, so we can't prove it's safe to remove.

The bottom line is: you shouldn't need to worry about the precise arcana of how the build and minification work in order to get guaranteed branch elimination, which is why `macroCondition` exists and why you should always put the code that you want eliminated into either it's consequent block or its alternate block.
